libusb/descriptor.c

```diff
@@ -1185,18 +1185,20 @@ int API_EXPORTED libusb_get_string_descriptor_ascii(libusb_device_handle *dev_ha
 		usbi_warn(HANDLE_CTX(dev_handle), "suspicious bLength %u for string descriptor (read %d)", str.desc.bLength, r);
 
 	di = 0;
```

Comment:
I would suggested dropping variable `di` since it is now 1:1 with `si` and only use for breaking the loop. Running the loop only while `si`/`di` is smaller than `length - 1` should be done by properly setting `si_max`

Reply:
There is no guarantee or requirement that the source and destination buffers are the same size.  Buffer overruns are disastrously common in C. It's safer and easier to reason about preventing them when you know the size of each buffer and guard each buffer with its corresponding size and separate index. Trying to cleverly reduce down to one index makes it harder to reason about.

Reply:
Thanks. My suggestion was actually to replace current `si_max `with someting like `Max(length -1,  (str.desc.bLength - sizeof(uint8_t) - sizeof(uint8_t)) / sizeof(uint16_t))` to express that we'll loop until we reach the first smallest boundary and also use only one loop variable instead of two.
I feel the code would look more "common" and easy to understand / modify, but I guess that is more a matter of taste.

Reply:
Ah, thanks for clarifying!

But you've just overrun the buffer again. :)  You want Min(), not Max(). :)  One might argue that this is again trying to be too clever, and leads to mistakes.

But with that corrected, I'd be ok with that formulation. How about you @tormodvolden ?  If so, I'll update this PR.

Reply:
Damn' ! Believe me: I clearly wanted to write `Min` ;-) Anyway, that is the only way my proposal make, so you autocorrected it, great.

Reply:
Yes, I like the Min() version too. Then it is clear from the start how long the for loop goes, without having to break out of it conditionally in the middle. And we can use `i` instead of `si` :)

Reply:
OK, see updated commit.